Add hipFile examples (#465)
amd/rocm-examples: Systems/hipFile/README.md

# hipFile Examples

This directory contains working examples of the hipFile API, grouped by what they demonstrate.

> Every program verifies its result with an FNV-1a hash and prints `OK …` on success.

Reviewer: ```suggestion
Every example verifies its result with an FNV-1a hash and prints `OK …` on success.
```

> Most examples move data through the GPU on hipFile's fast path, which opens files with `O_DIRECT`.

Reviewer: ```suggestion
Most examples move data through the GPU use the hipFile fastpath backend, which opens files with `O_DIRECT`.
```

> Running them therefore requires an AMD GPU supported by ROCm and source/destination paths on an

Reviewer: ```suggestion
Running them requires an AMD GPU supported by ROCm and source/destination paths on an
```

`O_DIRECT`-capable local filesystem (ext4 mounted `data=ordered`, or xfs). `O_DIRECT` is not a

> hipFile requirement — files can be opened without it and routed through the POSIX compat path (see

Reviewer: ```suggestion
hipFile requirement. Files can be opened without it and routed through the POSIX fallback backend (see
```

> [`basics/no_odirect_write`](basics/no_odirect_write)). Verify fast-path support with

Reviewer: ```suggestion
[`basics/no_odirect_write`](basics/no_odirect_write)). Verify fastpath support with
```

`/opt/rocm/bin/ais-check`. The [`api`](api) examples are the exception: they only query the
library and require neither a GPU nor an `O_DIRECT` filesystem.

## Prerequisites

- [ROCm](https://rocm.docs.amd.com/projects/HIP/en/latest/install/install.html) (at least version 7.x)
- hipFile (installed via the nightly `.deb` packages or built from source)
- CMake (at least version 3.21), or GNU Make

## Directory layout

| Directory | What's in it |
| --------- | ------------ |

> | [`api`](api) | Minimal examples of the non-I/O API — calls that query or configure the library (e.g. `get_version`). No `O_DIRECT` filesystem or file arguments needed. |

Reviewer: ```suggestion
| [`api`](api) | Minimal examples of the non-I/O API calls that query or configure the library (e.g. `get_version`). No `O_DIRECT` filesystem or file arguments needed. |
```

> | [`basics`](basics) | Small, single-purpose programs that each exercise one facet of the synchronous API: buffer registration, the `O_DIRECT` fast path vs. compat fallback, chunked reads, device-buffer offsets, sub-region writes, GPU memory types, and a full round trip. |

Reviewer: ```suggestion
| [`basics`](basics) | Small, single-purpose programs that each use one facet of the synchronous API: buffer registration, the fastpath backend, the fallback backend, chunked reads, device-buffer offsets, sub-region writes, GPU memory types, and a full round trip. |
```

| [`async`](async) | Examples of the asynchronous, stream-based API (`hipFileReadAsync` / `hipFileWriteAsync`), including single-stream, non-blocking-stream, and concurrent multi-stream round trips. |
| [`aiscp`](aiscp) | A standalone `cp`-like utility built on hipFile (`hipfile_aiscp SOURCE DEST`). |

> | [`common`](common) | Shared helpers used by `basics` and `async` (alignment math, pattern fill, FNV-1a hashing, file open/register). Not an example — compiled directly into each example that needs it. |

Reviewer: ```suggestion
| [`common`](common) | Shared helpers used by `basics` and `async` (alignment math, pattern fill, FNV-1a hashing, file open/register). This is not an example in and of itself, but is compiled directly into each example that requires it. |
```

## Building

### CMake (from the repository root)

```bash
cmake -DCMAKE_BUILD_TYPE=Release -S . -B build
cmake --build build --parallel
```

The binaries land under `build/bin/Systems/hipFile/`.

### CMake (Systems subtree only)

```bash
cd Systems
cmake -DCMAKE_BUILD_TYPE=Release -S . -B build
cmake --build build --parallel
```

### CMake (single example)

Each example directory is a self-contained CMake project. For example:

```bash
cd Systems/hipFile/basics/bufregister_write
cmake -S . -B build
cmake --build build
```

### Make

```bash
cd Systems
make -j $(nproc)
```

Or for a single category:

```bash
cd Systems/hipFile/basics
make -j $(nproc)
```

---

## `api`

> Minimal examples of the non-I/O parts of the hipFile API — calls that query or configure the

Reviewer: ```suggestion
Minimal examples of the non-I/O parts of the hipFile API. These are calls that query or configure the
```

library rather than move data through the GPU. These do **not** require an `O_DIRECT`-capable
filesystem or file arguments.

| Program | What it shows | Args |
| ------- | ------------- | ---- |
| [`get_version`](api/get_version) | Read the hipFile version both ways: the `HIPFILE_VERSION_*` header macros (compile-time) and `hipFileGetVersion()` (runtime). | none |

### Running

```bash
./hipfile_get_version
```

Prints the version from the header symbols and from the runtime call. No file or GPU memory is
touched.

---

## `basics`

> Small, single-purpose programs that each exercise one facet of the synchronous hipFile C API.

Reviewer: ```suggestion
Small, single-purpose programs that each demonstrate one facet of the synchronous hipFile C API.
```
